jplag/src/main/java/de/jplag/clustering/ClusteringResult.java

```diff
@@ -92,11 +89,33 @@ public static ClusteringResult<Integer> fromIntegerCollections(List<Collection<I
             for (int i = 0; i < clustering.size(); i++) {
                 double outWeightSum = percentagesOfSimilaritySums.getRowVector(i).getL1Norm();
                 double clusterCommunityStrength = percentagesOfSimilaritySums.getEntry(i, i) - outWeightSum * outWeightSum;
-                clusters.add(new Cluster<>(clustering.get(i), (float) clusterCommunityStrength));
+                float averageSimilarity = calculateAverageSimilarityFor(clustering.get(i), similarity);
+                clusters.add(new Cluster<>(clustering.get(i), (float) clusterCommunityStrength, averageSimilarity));
                 communityStrength += clusterCommunityStrength;
             }
         }
         return new ClusteringResult<>(clusters, communityStrength);
     }
 
+    private static float calculateAverageSimilarityFor(Collection<Integer> cluster, RealMatrix similarityMatrix) {
+        var sumOfSimilarities = 0f;
+        var submissionIndicesWithoutIndicesAlreadyProcessed = new ArrayList<>(List.copyOf(cluster));
+        for (Integer indexOfSubmission1 : cluster) {
+            for (Integer indexOfSubmission2 : submissionIndicesWithoutIndicesAlreadyProcessed) {
+                if (!Objects.equals(indexOfSubmission1, indexOfSubmission2)) {
+                    sumOfSimilarities += similarityMatrix.getEntry(indexOfSubmission1, indexOfSubmission2);
+                }
+            }
+            submissionIndicesWithoutIndicesAlreadyProcessed.remove(indexOfSubmission1); // remove indices we already processed from second list to avoid adding them anew unnecessary
+        }
```

Comment:
If the dimensions of `similarityMatrix` are roughly equivalent to `cluster.size()`, I would consider calculating the sum using a `RealMatrixVisitor` instead, like so:
```suggestion
        double sumOfSimilarities = similarityMatrix.walkInOptimizedOrder(new DefaultRealMatrixPreservingVisitor() {
            double sum;

            @Override
            public void visit(int row, int column, double value) {
                if (row > column) sum += value;
            }

            @Override
            public double end() {
                return sum;
            }
        });
```
If, on the other hand, `similarityMatrix` is big and sparse, I suggest using the standard `for`-loop to iterate over the indexes in `cluster`, because you are essentially emulating that.
```suggestion
        var sumOfSimilarities = 0f;
        for (int index1 = 1; index1 < cluster.size(); index1++) {
            for (int index2 = 0; index2 < index1; index2++) {
                sumOfSimilarities += similarityMatrix.getEntry(index1, index2);
            }
        }
```
Both approaches save you from having to test for any special cases or managing the `submissionIndicesWithoutIndicesAlreadyProcessed`.

Reply:
Hey Robin, thanks for your feedback! While this is not a fitting case for the `RealMatrixVisitor` I'm happy you told me about it, as I did not know it before.
Concerning changes to the loop: The streams approach is clever, I really like it. But I do share your concerns with readability. I do find your standard for-loop approach more precise and efficient - I will incorporate it.